Lib/shutil.py

```diff
@@ -606,37 +606,23 @@ def _rmtree_islink(st):
 
 # version vulnerable to race conditions
 def _rmtree_unsafe(path, onexc):
-    try:
-        with os.scandir(path) as scandir_it:
-            entries = list(scandir_it)
-    except FileNotFoundError:
-        return
-    except OSError as err:
-        onexc(os.scandir, path, err)
-        entries = []
-    for entry in entries:
-        fullname = entry.path
-        try:
-            is_dir = entry.is_dir(follow_symlinks=False)
-        except FileNotFoundError:
-            continue
-        except OSError:
-            is_dir = False
-
-        if is_dir and not entry.is_junction():
+    def onerror(err):
+        if not isinstance(err, FileNotFoundError):
+            onexc(os.scandir, err.filename, err)
+    results = os.walk(path, False, onerror, os._walk_symlinks_as_files)
+    for dirpath, dirnames, filenames in results:
+        # Add trailing slash to dirpath.
+        dirpath = os.path.join(dirpath, dirpath[:0])
```

Comment:
```suggestion
        dirpath = os.path.join(dirpath, "")
```
Am I missing something or is this a convoluted way to spell the empty string?

Also, is this basically a performance optimization so we can use `+` below instead of `os.path.join`?

Reply:
`dirpath` can be `bytes` here! I'll expand the comment.

Reply:
In fact, I'll just leave this optimisation out and save it for another PR. Clearer code is better for now.

Reply:
Oh thanks, I forgot about bytes paths! I think the optimization is fine with a comment, but agree it can wait for a later PR.